### PR TITLE
 Add reusable actions to create and/or verify a new project 

### PR DESCRIPTION
**Pull request by rmartin16**
## Changes
- Introduces a composite action to create a Briefcase project based on a GUI toolkit and optional template
- Introduces a workflow, `app-create-verify.yml`, for repos to run in CI that have impacts on project creation
  - Notably, `beeware/briefcase-template` and `beeware/briefcase` should implement this

BRIEFCASE_REPO: https://github.com/rmartin16/briefcase
BRIEFCASE_REF: gui-plugin-support
BRIEFCASE_TEMPLATE_REPO: https://github.com/rmartin16/briefcase-template
BRIEFCASE_TEMPLATE_REF: gui-plugin-support

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by rmartin16** [comment]
This really creates some chicken and egg problems....

**Comment by rmartin16** [comment]
So, I'm trying to get to a place where none of the downstream repos need to have project creation config or pre-rolled out apps for CI. I've removed the need for the config...but the rolled apps will be more difficult since they include specific `pyproject.toml` changes to build the app using the current ref of the repo you're trying to test.

This really seems quite ripe for a plugin that adds `template = ../../` to each app format table....

**Comment by freakboy3742** [comment]
> So, I'm trying to get to a place where none of the downstream repos need to have project creation config or pre-rolled out apps for CI. I've removed the need for the config...but the rolled apps will be more difficult since they include specific `pyproject.toml` changes to build the app using the current ref of the repo you're trying to test.
> 
> This really seems quite ripe for a plugin that adds `template = ../../` to each app format table....

I was literally about to break ground on this feature today. tl;dr `briefcase create -C template=...` so you can override individual app settings at runtime.

**Comment by rmartin16** [comment]
Status
- This is ready for at least an initial review
- We'll need https://github.com/beeware/briefcase/pull/1542 to ensure `briefcase create` is being called correctly....but the AppImage test in `app-build-verify` demonstrates what this'll look like.

File: .github/workflows/app-create-verify.yml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        assert pyproject_toml.is_file()
+        pprint(toml.load(pyproject_toml))
+
+    - name: Install Dependencies (Ubuntu)
+      if: startsWith(inputs.runner-os, 'ubuntu')
+      run: sudo apt update -y && sudo apt install -y --no-install-recommends libgirepository1.0-dev
+
+    - name: Test Project Build
+      working-directory: ${{ steps.create.outputs.project-path }}
+      run: briefcase build
```
Comment (rmartin16) [inline .github/workflows/app-create-verify.yml:122]: This was mentioned previously so I included it....but it introduces a bunch of complications and both `beeware/briefcase-template` and `beeware/briefcase` run `app-build-verify.yml` which will perform app builds to a much more thorough extent. So, I'm leaning towards taking this out.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This mostly looks good; a couple of questions and clarifications inline.

File: .github/workflows/app-build-verify.yml [hunk 2]
```diff
@@ -254,28 +257,30 @@ jobs:
...
       if: >
         startsWith(inputs.runner-os, 'ubuntu')
         && contains(fromJSON('["Linux"]'), inputs.target-platform)
         && contains(fromJSON('["AppImage"]'), inputs.target-format)
+      working-directory: ${{ steps.create.outputs.project-path }}
       run: |
-        cd tests/apps/verify-${{ inputs.framework }}
-        briefcase create linux AppImage
+        briefcase create linux AppImage \
+          ${{ steps.output-format.outputs.template-override }} \
```
[the remark this answers is not in the document]
Comment (rmartin16) [reply .github/workflows/app-build-verify.yml:268]: They all need it; I will add them once https://github.com/beeware/briefcase/pull/1542 is merged. If I specify `--config` now, everything will fail so I only did it for AppImage since it always fails without `--config`.

File: .github/workflows/app-create-verify.yml (hunk 1 above)
Comment (rmartin16) [reply .github/workflows/app-create-verify.yml:122]: > the issue is that the test is running on Py3.12, which is completely fine for generating the template, but not for _building_ on Linux because of the system python dependency?

Right; running `briefcase build` basically imposes all of the restrictions of building that are already accommodated in `app-build-verify` and both these repos already run.

File: .github/workflows/app-build-verify.yml [hunk 3]
```diff
@@ -254,28 +257,30 @@ jobs:
...
+          --config 'requires=["toga-gtk==0.3.1", "PyGobject<3.46.0"]'
```
[the remark this answers is not in the document]
Comment (rmartin16) [reply .github/workflows/app-build-verify.yml:269]: Yeah; good point. From going through all the templates today, AppImage is only one I saw with a special additional configuration requirement beyond specifying `template=../../`.

I think for where to declare a build config override like this....I'm leaning towards here in `app-build-verify`....otherwise, any repo that needed to invoke a build with special requirements would need to be aware of such requirements....I'll consider it more though.

File: .github/actions/install-briefcase/action.yml [hunk 4]
```diff
@@ -58,10 +58,8 @@ runs:
         fi
 
         # Expose repo and version via outputs
-        echo "Briefcase Repo: ${BRIEFCASE_REPO}"
-        echo "Briefcase Ref: ${BRIEFCASE_REF}"
-        echo "repo=${BRIEFCASE_REPO}" >> ${GITHUB_OUTPUT}
-        echo "ref=${BRIEFCASE_REF}" >> ${GITHUB_OUTPUT}
+        echo "repo=${BRIEFCASE_REPO}" | tee -a ${GITHUB_OUTPUT}
+        echo "ref=${BRIEFCASE_REF}" | tee -a ${GITHUB_OUTPUT}
```
[the remark this answers is not in the document]
Comment (rmartin16) [reply .github/actions/install-briefcase/action.yml:62]: I know right....it finally dawned on me after looking at enough repetitions of `echo` to `stdout` and then `echo` to file.

File: .github/workflows/app-build-verify.yml [hunk 5]
```diff
@@ -254,29 +259,38 @@ jobs:
...
+        # PyGObject>=3.46.0 requires a version of glibc that isn't available in manylinux images;
+        # so, the version will need to be constrained to successfully build an AppImage with Toga.
+        # Furthermore, Toga>0.3.1 requires PyGObject>=3.46.0 so its version is constrained as well.
+        if [ "${{ startsWith(inputs.framework, 'toga') }}" = "true" ]; then
+          CONFIG_OVERRIDE_REQUIRES='requires=["toga-gtk==0.3.1", "PyGobject<3.46.0"]'
+        fi
+
+        briefcase create linux AppImage \
+          ${{ steps.output-format.outputs.template-override }} \
+          --config $(printf '%s' $CONFIG_OVERRIDE_REQUIRES)
```
Comment (rmartin16) [inline .github/workflows/app-build-verify.yml:278]: I tried a lot of iterations of escape sequences to handle this more straightforwardly....but none of them were working for me and I landed here. Definitely open to improvement.

File: .github/workflows/app-build-verify.yml [hunk 6]
```diff
@@ -100,144 +101,148 @@ jobs:
           ${{ steps.dirs.outputs.pip-cache-dir }}
           ${{ steps.dirs.outputs.docker-cache-dir }}
```
Comment (rmartin16) [inline .github/workflows/app-build-verify.yml:102]: We can probably be more thoughtful in this caching.

For one, `beeware/.github` is [hitting](https://github.com/beeware/.github/actions/caches?query=sort%3Asize-desc) the 10GB cache storage limit....and Briefcase has several of these [caches](https://github.com/beeware/briefcase/actions/caches?query=sort%3Asize-desc) that are all nearly the same size....implying they are likely duplicates.

## tl;dr
I've provided rationale to only cache the Briefcase directory here (and `pip` using `actions/setup-python`).

Additionally, the cache key will be updated to remove inclusion of the GUI toolkit and add the current month and year.

## Caching `pip`
We don't need to manage the `pip` cache; `actions/setup-python` can handle that for us; for instance:
```yaml
  - name: Setup Python
    uses: actions/setup-python@v4.7.1
    with:
      python-version: ${{ inputs.python-version }}
      cache: pip
      cache-dependency-path: |
        **/setup.cfg
        **/pyproject.toml
```

## Caching `~/cookiecutters`
We're caching the downloaded cookiecutters. On its face, this sounds like a benefit....but the cache doesn't just materialize in the runner...it has to be downloaded. So, what's the difference in downloading the cache from GitHub or using `git` to download the cookiecutter on-demand from GitHub. These are effectively equivalent actions but one doesn't count against the 10GB cache limit. Additionally, only the macOS app template repo really has any heft to it anyway...so the benefit here is pretty marginal regardless.

## Stale caches
Third, when a workflow job has a cache hit, it doesn't upload any changes committed to that cache after the run. Looking at Briefcase's caches, some of them are 9 months old. That means we're downloading a cache created 9 months ago and using that as a base to run CI.

This is probably negating at least some of the desired benefits of caching here since a bunch of old data is (a) going to need to be updated during CI anyway or (b) degrades the fidelity of the testing process because we're using a Briefcase directory that was created a long time ago. For instance, right now, Briefcase CI has to re-download and install Android Commandline Tools 9.0 every run because it isn't in the current version of the cache.

To mitigate this, I can add the month and year to the cache key so they are evicted regularly. Unfortunately, this will create effectively duplicate entries among the caches over time. In turn, that means the repo will eventually always be up against its 10GB limit and GitHub will evict the least recently used entry. In general, this is bad because entries that are useful but not used often may be evicted unnecessarily. However, I don't think we really use caching like that anywhere; basically all of the caches are used in every CI run....so, this probably isn't too big of a deal.

An alternative approach might be to fingerprint the Briefcase directory; perhaps a recursive hash of all the files. This hash would then become a part of the cache key ensuring a new cache is created anytime something in the directory changes. This seems a bit over-engineered (and probably slow)....and I haven't completely thought through if files might change in the Briefcase directory as part of normal operation.

## Docker caching
We're including Docker caches except for on Linux (because GitHub doesn't support it on Linux)...but Docker is only used on Linux in CI. That said, even if we could cache Docker, the manylinux images start at over 1GB....so, it probably wouldn't even be reasonable to put them in GH cache given the 10GB limit.

That said, GitHub provides an unlimited storage amount of images published to `ghcr.io`. So, if we did want to entertain Docker caching, we could simply push the final image to GHCR and load it prior to running CI.

I'm not going to attempt that in this PR, though.

## Defining the cache key
This is the current key:
```
briefcase-${{ runner.os }}-${{ inputs.repository }}-${{ inputs.framework }}-${{ inputs.target-platform }}-${{ inputs.target-format }}
```
It basically ensures a new cache is created for any permutation of invoking this workflow. However, which of these inputs really contribute to a actually unique cache?

- `runner.os`: keep
  - Caches will certainly vary by OS
- `inputs.repository`: keep
  - Cache will be different for Android on Linux than AppImage
  - This is where updating an existing cache could be particularly useful....I'm imagining that the caches from these different runs on the same OS are all merged together to create a "super-cache"....but that starts to become really particular to this use-case
- `inputs.framework`: remove
  - Given this is really just about the Briefcase directory, it is independent of the GUI toolkit being used
- `inputs.target-platform`: keep
  - Targeting Android vs Linux is going to create different caches
- `inputs.target-format`: keep
  - Since the cache is already limited to a target platform, this is really whether the Briefcase directory is different for different formats for the same platform.
  - The Briefcase directory is generally consistent per target platform...except on Linux. And there's nothing really preventing the directory not changing further in the future for each target platform.

Comment (rmartin16) [reply .github/workflows/app-build-verify.yml]: > - `inputs.framework`: remove
>     * Given this is really just about the Briefcase directory, it is independent of the GUI toolkit being used

Regarding the cache key....while the GUI toolkit does not explicitly change the contents of the Briefcase data directory, it does so implicitly through the implementation of the `app-build-verify` workflow itself.

For instance, when Briefcase's CI runs `app-build-verify` for Toga on Linux and another run for PySide6 on Linux, both of these will generate the same cache key. However, the Toga run will build an app for Android while the PySide6 run will not. Therefore, the data directories will be different but only one of them will be allowed to push the cache up for later runs.

Therefore, the GUI toolkit should remain a part of the cache key after all.

Comment (rmartin16) [reply .github/workflows/app-build-verify.yml]: > For instance, right now, Briefcase CI has to re-download and install Android Commandline Tools 9.0 every run because it isn't in the current version of the cache.

While it is true that Briefcase has to download Commandline Tools 9.0 for each workflow run, it is not because the cache is stale. It is because CI is using the Android SDK that already exists on the runner; the pre-installed SDK has a version of Commandline Tools installed as `latest` so Briefcase is using that version to then install version 9.0.

This could be changed by forcing Briefcase to ignore this Android SDK and set up its own SDK that would also even be cached. However, I would expect this to become untenable given the potential size of the SDK. This will become even more so the case once I introduce `briefcase run` in to `app-build-verify` where the Android system image will be downloaded.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -19,13 +19,13 @@ env:
...
-    name: Pre-commit checks
+    name: Pre-commit Checks
     uses: ./.github/workflows/pre-commit-run.yml
     with:
-      pre-commit-source: "pre-commit"
+      pre-commit-source: pre-commit
 
   test-docs-build-verify:
-    name: Test Verify Docs Build
+    name: Verify Docs Build Check
```
Comment (freakboy3742) [inline .github/workflows/ci.yml:28]: Minor Nit: The task is verifying the docs build; verifying a check is... a bit meta.
```suggestion
    name: Verify Docs Build
```

File: .github/workflows/app-create-verify.yml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        repository: rmartin16/.github-beeware  # !!!! TODO:PR: REPLACE ME WITH beeware/.github !!!!
+        ref: create-new-project  # !!!! TODO:PR: REMOVE ME !!!!
+        path: beeware-.github
+
+    - name: Checkout beeware/briefcase-template
+      uses: actions/checkout@v4.1.1
+      with:
+        repository: rmartin16/briefcase-template  # !!!!! TODO:PR: REPLACE ME WITH beeware/briefcase-template !!!!!
+        ref: gui-plugin-support  # !!!!! TODO:PR REMOVE ME !!!!!
+        path: briefcase-template
```
Comment (rmartin16) [inline .github/workflows/app-create-verify.yml:62]: It isn't clear why I added this....but it definitely isn't necessary now. It should be removed when the updates are made prior to merging.

Comment (rmartin16) [reply .github/workflows/app-create-verify.yml:62]: Ohh nvm, i see now. This is needed to install the requirements for linting the project. So, it should stick around.

File: .github/workflows/app-build-verify.yml (hunk 6 above)
Comment (freakboy3742) [reply .github/workflows/app-build-verify.yml]: Agreed that pip and cookiecutter caches can be optimised in the way you describe. Also agreed that there's an opportunity for Docker image caching, but we don't need to solve that problem here.

I'm fairly certain the rest of your caching argument is sound; and AFAICT the implementation essentially comes down to the cache key - and the worst case is "the data you need isn't cached and needs to be downloaded on demand", so AFAICT we shouldn't see *failures* as a result of any caching issue, just longer test times. We can iterate on these over time if we need to.

Some caching related thoughts/ideas. Unless you're feeling especially inspired, I don't think we need to tackle these as part of this PR. Even if we turn *off* caching, all we see is slightly slower test runs; I'm only flagging these since the topic has come up.

1. Given the inconsistency of interpretation in `input.target-format`, I wonder if a partial solution is to make the cache key itself platform dependent. That would allow macOS to use the same key for all formats, but allow AppImage and Flatpak to have different keys.

2. Another possible approach would be to fracture the cache into smaller parts, and pull it on demand. Rather than a single "Briefcase" cache, cache the support packages, android toolchain, java toolchain, etc, and request the specific cache keys that are needed. Harder to manage, but less likely to end up with competing cache keys.

3. We've been deliberately avoiding the pre-installed tools on the CI machines... but I wonder if it might be worth relaxing that (or, at least, relaxing them as an option). The two tools that actually benefit from caching are Android SDK and Java - the remaining tools are sufficiently small that a cache miss isn't *that* big a deal. The versions provided on-machine are knowable and reliable; Java 17 is available on all images; and the Ubuntu-latest  image [ships with command line tools version 9](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md) - so we're not actually gaining anything by explicitly downloading those tools, other than exercising the download mechanism. If the default Android SDK version were to change, that would be an interesting quandary - but it would also be a useful test that we're tracking the behavior of the Android toolchain. 

If we take the Android and Java components out of the toolchain, we also remove a *lot* of our machine needs - the next biggest chunk would be the support packages, and those are pretty safe to cache on a per-OS basis.

Comment (rmartin16) [reply .github/workflows/app-build-verify.yml]: > so AFAICT we shouldn't see failures as a result of any caching issue, just longer test times.

My biggest concern is undetected/unreported issues because CI is running with a stale/old cache....although, these seem probably rather unlikely. Nonetheless, I'm still planning to implement a quick job that will periodically delete all the caches so they can be re-built.

> Unless you're feeling especially inspired, I don't think we need to tackle these as part of this PR.

Beyond what I've already done, I'd rather save further cache tweaking for the future. These are definitely some good considerations.

File: .github/workflows/app-build-verify.yml [hunk 9]
```diff
@@ -63,181 +63,191 @@ jobs:
...
       if: >
         startsWith(inputs.runner-os, 'ubuntu')
-        && !contains(fromJSON('["PySide2"]'), inputs.framework)
         && contains(fromJSON('["", "Linux"]'), inputs.target-platform)
         && contains(fromJSON('["", "system"]'), inputs.target-format)
+      working-directory: ${{ steps.create.outputs.project-path }}
       run: |
-        cd tests/apps/verify-${{ inputs.framework }}
-        briefcase create linux system --target fedora:37
+        briefcase create linux system --target fedora:37 \
```
Comment (freakboy3742) [inline .github/workflows/app-build-verify.yml:237]: Doesn't need to be fixed in this PR, but flagging that Fedora 37 is EOL in 2 weeks, so we're likely going to need to bump this to 38 or 39 soon.

File: .github/workflows/app-build-verify.yml [hunk 10]
```diff
@@ -254,29 +264,39 @@ jobs:
...
         startsWith(inputs.runner-os, 'ubuntu')
         && contains(fromJSON('["", "Linux"]'), inputs.target-platform)
         && contains(fromJSON('["", "Flatpak"]'), inputs.target-format)
         && contains(fromJSON('["Toga", "PyGame"]'), inputs.framework)
+      working-directory: ${{ steps.create.outputs.project-path }}
       run: |
-        sudo apt-get update -y
-        sudo apt-get install -y flatpak flatpak-builder
+        sudo apt update -y
+        sudo apt install -y --no-install-recommends flatpak flatpak-builder elfutils
```
Comment (freakboy3742) [inline .github/workflows/app-build-verify.yml:296]: What is elfutils for? Why the new requirement?

Comment (rmartin16) [reply .github/workflows/app-build-verify.yml:296]: I added `--no-install-recommends` to avoid all the stuff we don't need that `apt` installs by default.

As for `elfutils`, it's only a recommended package for `flatpak-builder` but it's required because it contains `eu-strip` which `flatpak-builder` needs.

**Review by freakboy3742 — APPROVED** [review]
Marking as approved, conditional on changes that are needed as part of the [beeware/briefcase#1524 landing strategy](https://github.com/beeware/briefcase/pull/1524#pullrequestreview-1736118477)

File: .github/workflows/app-build-verify.yml (hunk 9 above)
Comment (rmartin16) [reply .github/workflows/app-build-verify.yml:237]: 👍🏼 I should take care of this in my next round of PRs. I think some of the versions in the templates (e.g. for `org.kde.Platform`) need bumps as well.

**Review by freakboy3742 — APPROVED** [review]
Approved for final ship.

File: .github/workflows/app-build-verify.yml (hunk 6 above)
Comment (rmartin16) [reply .github/workflows/app-build-verify.yml]: > 3\. We've been deliberately avoiding the pre-installed tools on the CI machines... but I wonder if it might be worth relaxing that (or, at least, relaxing them as an option). The two tools that actually benefit from caching are Android SDK and Java - the remaining tools are sufficiently small that a cache miss isn't _that_ big a deal. The versions provided on-machine are knowable and reliable; Java 17 is available on all images; and the Ubuntu-latest  image [ships with command line tools version 9](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md) - so we're not actually gaining anything by explicitly downloading those tools, other than exercising the download mechanism. If the default Android SDK version were to change, that would be an interesting quandary - but it would also be a useful test that we're tracking the behavior of the Android toolchain.

Using the Java 17 that GitHub that provides will be trivial; it's location is in `$JAVA_HOME_17_X64`.

We actually already use the GitHub Android SDK. And it does contain Command-Line Tools 9.0 already....however, the tools are installed as `latest` and not `9.0`. I could update Briefcase to use `latest` if it is indeed 9.0....but in reality this download doesn't take long at all. Even downloading the system image to run the emulator takes <30s.

At any rate, using the GitHub Java 17 prevents Briefcase from using the data directory at all.
